### PR TITLE
determining executable activities

### PR DESCRIPTION
**Pull request by Noel-Bastubbe**
Description: 
- adds the required data types to determine executable activities
- adds the ability to determine executable activities
- adds tests to determining executable activities

## PR checklist

- [x] dev-branch has been merged into local branch to resolve conflicts
- [x] the application has been manually tested after merge
- [ ] another dev reviewed and approved
- [ ] if feature-Branch: Teams PO has approved (show via e.g. screenshots/screencapture/live demo)

### CODE REVIEW

File: planner/types.ts [hunk 1]
```diff
@@ -0,0 +1,161 @@
+export class Activity {
+    name: string;
+    duration: number;
+    NoP: number;
+    role: Role;
+    input: IOSet[];
+    output: IOSet[];
+
+    public constructor(name: string, duration: number, NoP: number, role: Role, input: IOSet[], output: IOSet[]) {
```
Comment (AnjoSs) [inline planner/types.ts:9]: You could already add default values for the duration (1) and the NoP (1) and maybe null for the role.

File: planner/types.ts [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+    first: Instance;
+    second: Instance;
+
+    public constructor(first: Instance, second: Instance) {
+        this.first = first;
+        this.second = second;
+    }
+}
+
+export class ProjectState {
```
Comment (AnjoSs) [inline planner/types.ts:110]:
```suggestion
export class ExecutionState {
```

File: planner/types.ts [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+    name: string;
+    dataclass: Dataclass;
+
+    public constructor(name: string, dataclass: Dataclass) {
+        this.name = name;
+        this.dataclass = dataclass;
+    }
+}
+
+export class Instance {
```
Comment (AnjoSs) [inline planner/types.ts:84]:
```suggestion
export class DataObject {
```

Comment (AnjoSs) [inline planner/types.ts:84]:
```suggestion
export class DataObject {
```

File: planner/types.ts [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+
+export class Dataclass {
+    name: string;
+
+    public constructor(name: string) {
+        this.name = name;
+    }
+}
+
+export class State {
```
Comment (AnjoSs) [inline planner/types.ts:74]: You could call it `DataObjectNode` with the attributes `class`, `state`, and `isList` to make sure it maps to the nodes in the fragments.

File: planner/types.ts [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+}
+
+export class IOSet {
+    set: Array<{ dataObject: State, isList: boolean }>;
+
+    public constructor(set: Array<{ dataObject: State, isList: boolean }>) {
+        this.set = set;
+    }
+
+    public satisfiedBy(currentState: State[]): boolean {
```
Comment (AnjoSs) [inline planner/types.ts:30]: This should be the execution state I guess

File: planner/types.ts [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+        let satisfiedBy = true;
+        this.set.forEach(function (State) {
+                if (!currentState.includes(State.dataObject)) {
+                    satisfiedBy = false;
```
Comment (AnjoSs) [inline planner/types.ts:34]:
```suggestion
                    return false;
```

File: planner/types.ts [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+                }
+            }
+        );
+        return satisfiedBy;
```
Comment (AnjoSs) [inline planner/types.ts:38]:
```suggestion
        return true;
```

File: planner/types.ts [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+export class IOSet {
+    set: Array<{ dataObject: State, isList: boolean }>;
+
+    public constructor(set: Array<{ dataObject: State, isList: boolean }>) {
+        this.set = set;
+    }
+
+    public satisfiedBy(currentState: State[]): boolean {
+        let satisfiedBy = true;
+        this.set.forEach(function (State) {
```
Comment (AnjoSs) [inline planner/types.ts:32]:
```suggestion
        this.set.forEach(function (dataObjectNode) {
```

File: planner/types.ts [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+        this.output = output;
+    }
+
+    public isExecutable(currentState: State[], ressources: Resource[]): boolean {
+        return this.input.some(inputSet => inputSet.satisfiedBy(currentState)) && ressources.some(resource => resource.satisfies(this.role, this.NoP));
+    }
+}
+
+export class IOSet {
+    set: Array<{ dataObject: State, isList: boolean }>;
```
Comment (AnjoSs) [inline planner/types.ts:24]: Use DataObjectNodes here instead

File: planner/types.ts [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+    public constructor(name: string, duration: number, NoP: number, role: Role, input: IOSet[], output: IOSet[]) {
+        this.name = name;
+        this.duration = duration;
+        this.NoP = NoP;
+        this.role = role;
+        this.input = input;
+        this.output = output;
+    }
+
+    public isExecutable(currentState: State[], ressources: Resource[]): boolean {
```
Comment (AnjoSs) [inline planner/types.ts:18]:
```suggestion
    public isExecutable(currentState: ExecutionState[], ressources: Resource[]): boolean {
```

File: planner/types.ts [hunk 11]
```diff
@@ -0,0 +1,161 @@
+export class Activity {
+    name: string;
+    duration: number;
+    NoP: number;
+    role: Role;
+    input: IOSet[];
+    output: IOSet[];
```
Comment (AnjoSs) [inline planner/types.ts:7]:
```suggestion
    inputSets: IOSet[];
    outputSets: IOSet[];
```

File: planner/types.ts [hunk 12]
```diff
@@ -0,0 +1,161 @@
+export class Activity {
```
[the remark this answers is not in the document]
Comment (AnjoSs) [reply planner/types.ts:1]: Or at least group the classes into something like activity and IOsets, data object-related stuff, and execution state, and objectives.

**Review by AnjoSs — DISMISSED** [review]
Looks great! I left some comments regarding the naming of your classes. I hope my suggestions make it a bit clearer. Maybe you have even better ideas.

**Review by Firsttryy — APPROVED** [review]
LGTM

**Review by Firsttryy — APPROVED** [review]
LGTM

**Review by simonmichaelnowack — APPROVED** [review]

**Pull request merged**